src/track/beats.cpp

```diff
@@ -661,6 +715,210 @@ std::optional<BeatsPointer> Beats::tryScale(BpmScale scale) const {
             m_subVersion));
 }
 
+std::optional<BeatsPointer> Beats::tryAdjustTempo(
+        audio::FramePos position, TempoAdjustment adjustment) const {
+    auto markers = m_markers;
+
+    // Retrieve the Beat Marker before the current position.
+    //
+    // Hence, we first get the marker *after* the position, then decrement the
+    // iterator.
+    // If there is no marker before the current position, the marker after the
+    // position can be used because the beats before the first marker are
+    // interpolated using the same beat length as the beats after the first
+    // marker.
+    auto markerIt = std::upper_bound(markers.begin(),
+            markers.end(),
+            position,
+            isFramePosLessThanBeatMarker);
+    if (markerIt != markers.begin() && position < m_lastMarkerPosition) {
+        markerIt--;
+    }
+
+    // At this point, `markerIt` points to the beat marker that we need to
+    // modify (or `cend()`, which means that we need to modify the BPM of the
+    // last marker).
+    Bpm lastMarkerBpm = m_lastMarkerBpm;
+    if (markerIt == markers.end()) {
+        lastMarkerBpm += (adjustment == TempoAdjustment::Faster) ? kBpmAdjustStep : -kBpmAdjustStep;
+        if (!lastMarkerBpm.isValid()) {
+            qWarning() << "Beats: Tempo adjustment would result in invalid Bpm!";
+            return std::nullopt;
+        }
+    } else {
+        const auto marker = *markerIt;
+        const int adjustedBeatsTillNextMarker = marker.beatsTillNextMarker() +
+                ((adjustment == TempoAdjustment::Faster)
+                                ? marker.beatsPerBar()
+                                : -marker.beatsPerBar());
+
+        if (adjustedBeatsTillNextMarker < marker.beatsPerBar()) {
+            qWarning() << "Beats: Tempo adjustment would result in a marker "
+                          "with less than the minimum beats in one bar!";
+            return std::nullopt;
+        }
+
+        markerIt = markers.erase(markerIt);
```

Comment:
Does erase shift (copy) all elements of the vector to close the gap? 
And happens it in the line below to revert it? This seems to be a topic for optimization. 
Can we just edit the original maker in place?

Reply:
Ok, it sounds like this is not necessary.